diff --git a/rts/eyrie.version b/rts/eyrie.version
index ac4ddd0..ffa253c 100644
--- a/rts/eyrie.version
+++ b/rts/eyrie.version
@@ -1 +1 @@
-a8be826
+d566dd85
diff --git a/scripts/init.sh b/scripts/init.sh
index 9c434e5..423ef19 100755
--- a/scripts/init.sh
+++ b/scripts/init.sh
@@ -18,7 +18,7 @@ init_runtime() {
 
     case $1 in
       eyrie)
-        repo_url=https://github.com/keystone-enclave/keystone-runtime
+        repo_url="-b dev-rv32 https://github.com/thuchoang90/keystone-runtime"
         ;;
       sel4)
         repo_url=https://github.com/keystone-enclave/keystone-sel4
